django/thunderstore/repository/api/experimental/serializers.py

```diff
@@ -95,3 +98,104 @@ class Meta:
             "has_nsfw_content",
             "categories",
         )
+
+
+class PackageUploadAuthorNameField(serializers.SlugRelatedField):
+    """Package upload's author name metadata field."""
+
+    def __init__(self, *args, **kwargs):
+        kwargs["slug_field"] = "name"
+        super().__init__(*args, **kwargs)
+
+    def get_queryset(self):
+        return UploaderIdentity.objects.filter(
+            members__user=self.context["request"].user,
+        )
+
+
+class PackageUploadCategoriesField(serializers.RelatedField):
+    """Package upload's categories metadata field."""
+
+    def get_queryset(self):
+        return PackageCategory.objects.filter(
+            community=self.context["request"].community,
+        )
+
+    def to_representation(self, value):
+        return [c.slug for c in value]
+
+    def to_internal_value(self, data):
+        if not isinstance(data, list):
+            raise serializers.ValidationError("Not a list")
+
+        out = []
+        for category_slug in data:
+            category = self.get_queryset().filter(slug=category_slug).first()
+            if not category:
+                raise serializers.ValidationError(
+                    f"'{category_slug}' category not found",
+                )
+            out.append(category)
+        return out
+
+
+class DictSerializer(serializers.Serializer):
+    def create(self, validated_data):
+        return validated_data
+
+
+class JSONToSerializerField(serializers.Field):
+    """Parses a JSON string and passes the data to a DictSerializer subclass."""
+
+    def __init__(self, serializer, *args, **kwargs):
+        self._serializer = serializer
+        super().__init__(*args, **kwargs)
+
+    def to_internal_value(self, data):
+        try:
+            serializer = self._serializer(data=json.loads(data), context=self.context)
+        except json.JSONDecodeError:
+            raise serializers.ValidationError("Invalid JSON")
+        serializer.is_valid(raise_exception=True)
+        return serializer.save()
```

Comment:
This seems prone for mistakes, e.g. if someone uses this field with a model serializer rather than a dict serializer. If we're calling the `.save` method, we better restrict this class to only work with `DictSerializer` or other non-db-modifying serialiszers.